python_ta/debug/accumulation_table.py

```diff
@@ -0,0 +1,155 @@
+
+    Private instance attributes:
+        loop_accumulators: a mapping between the accumulation variables
+            and their values during each iteration
+        loop_var_name: the name of the loop variable
+        loop_var_val: the values of the loop variable during each iteration
+        _loop_lineno: the line number of the for loop
+
+    """
+
+    loop_accumulators: dict[str, list]
+    loop_var_name: str
+    loop_var_val: list
+    _loop_lineno: int
+
+    def __init__(self, accumulation_names: list[str]) -> None:
+        self.loop_accumulators = {accumulator: [] for accumulator in accumulation_names}
+        self.loop_var_name = ""
+        self.loop_var_val = []
+        self._loop_lineno = 0
+
+    def _record_iteration(
+        self, accumulator_values: list, loop_variable_value: Any, frame: types.FrameType
+    ) -> None:
+        """Record the values of the accumulator variables and loop variable of an iteration"""
+        if len(self.loop_var_val) > 0:
+            self.loop_var_val.append(loop_variable_value)
+            for index, accumulator in enumerate(self.loop_accumulators):
+                self.loop_accumulators[accumulator].append(accumulator_values[index])
+        else:
+            # zeroth iteration
+            self.loop_var_val.append("N/A")
+            for name in self.loop_accumulators:
+                if name in frame.f_locals:
+                    self.loop_accumulators[name] = [frame.f_locals[name]]
+                else:
+                    raise NameError
+
+    def _create_iteration_dict(self) -> dict:
+        """Return a dictionary that maps each accumulator
+        and loop variable to its respective value during each iteration
+        """
+        return {
+            "iteration": list(range(len(self.loop_var_val))),
+            "loop variable (" + self.loop_var_name + ")": self.loop_var_val,
+            **self.loop_accumulators,
+        }
+
+    def _tabulate_data(self) -> None:
+        """Print the values of the accumulator and loop variables into a table"""
+        iteration_dict = self._create_iteration_dict()
+        print(
+            tabulate.tabulate(
+                iteration_dict, headers="keys", colalign=(*["left"] * len(iteration_dict),)
+            )
+        )
+
+    def _trace_loop(self, frame: types.FrameType, event: str, arg: Any) -> None:
+        """Trace through the for loop and store the values of the
+        accumulators and loop variable during each iteration
+        """
+        if event == "line" and frame.f_lineno == self._loop_lineno:
+            local_vars = frame.f_locals
+            curr_vals = []
+            for accumulator in self.loop_accumulators:
+                if accumulator in local_vars:
+                    curr_vals.append(local_vars[accumulator])
+                else:
+                    raise NameError
+
+            if self.loop_var_name in local_vars and len(self.loop_var_val) > 0:
```

Comment:
This check isn't necessary. It's okay to have `self._record_iteration` execute every time the loop line is reached, including the very first time ("iteration 0"). This means you do not have to call `self._record_iteration` in the `_setup_table` method, just here.